skbio/tree/_dists.py

```diff
@@ -0,0 +1,395 @@
+        Whether to include single-taxon biparitions (terminal branches) in the
+        calculation. Default is True.
+
+    Returns
+    -------
+    DistanceMatrix
+        Matrix of weighted Robinson-Foulds distances or variants.
+
+    See Also
+    --------
+    TreeNode.compare_wrfd
+
+    Notes
+    -----
+    The weighted Robinson-Foulds (wRF) distance [1]_ is the sum of differences of
+    branch lengths of matching bipartitions between a pair of trees.
+
+    This function is equivalent to :meth:`TreeNode.compare_wrfd` for two trees. Refer
+    to the latter for details about the metric and its variants, and the parameter
+    settings for calculating them. However, the current function extends the operation
+    to an arbitrary number of trees and returns a distance matrix for them.
+
+    A restriction of the current function compared to ``compare_wrfd`` is that
+    ``metric`` must be symmetric (i.e., :math:`d(x, y) = d(y, x)`), and equals zero
+    from a vector to itself (i.e., :math:`d(x, x) = 0`). It does not have to suffice
+    non-negativity or triangle inequality though.
+
+    This function is optimized for calculation based on taxa shared across all trees.
+    One can instead set ``pairwise`` to True to calculate based on taxa shared between
+    each pair of trees, which is however less efficient as bipartitions need to be
+    re-inferred during each comparison.
+
+    References
+    ----------
+    .. [1] Robinson, D. F., & Foulds, L. R. (1979) Comparison of weighted labelled
+        trees. In Combinatorial Mathematics VI: Proceedings of the Sixth Australian
+        Conference on Combinatorial Mathematics, Armidale, Australia (pp. 119-126).
+
+    Examples
+    --------
+    >>> trees = [TreeNode.read([x]) for x in (
+    ...     "((a:1,b:2):1,c:4,((d:4,e:5):2,f:6):1);",
+    ...     "((a:3,(b:2,c:2):1):3,d:8,(e:5,f:6):2);",
+    ...     "((a:1,c:6):2,(b:3,(d:2,e:3):1):2,f:7);",
+    ... )]
+    >>> dm = wrf_dists(trees, ids=list("ABC"))
+    >>> print(dm)
+    3x3 distance matrix
+    IDs:
+    'A', 'B', 'C'
+    Data:
+    [[  0.  16.  15.]
+     [ 16.   0.  27.]
+     [ 15.  27.   0.]]
+
+    """
+    _check_ids(trees, ids)
+    if pairwise:
+        metric = partial(
+            TreeNode.compare_wrfd,
+            metric=metric,
+            rooted=rooted,
+            include_single=include_single,
+        )
+        return DistanceMatrix.from_iterable(
+            trees, metric=metric, keys=ids, validate=False
+        )
+
+    metric, half = _check_dist_metric(metric)
+
+    method = "subsets" if rooted else "biparts"
+    withins = _withins(trees)
+    maps = [
+        getattr(x, method)(within=y, include_single=include_single, map_to_length=True)
+        for x, y in zip(trees, withins)
+    ]
+    gets = [x.get for x in maps]
+    sets = [frozenset(x) for x in maps]
+
+    result = np.zeros((n_trees := len(trees), n_trees))
+    for i, j in combinations(range(n_trees), 2):
```

Comment:
Same comment here, it seems like there is a missed opportunity to funnel this work all through the same implementation.